stan/math/rev/fun/columns_dot_self.hpp

```diff
@@ -16,14 +16,23 @@ namespace math {
  * @tparam Mat Type of the matrix.
  * @param x Matrix.
  */
-template <typename Mat, require_eigen_vt<is_var, Mat>* = nullptr>
-inline Eigen::Matrix<var, 1, Mat::ColsAtCompileTime> columns_dot_self(
-    const Mat& x) {
-  Eigen::Matrix<var, 1, Mat::ColsAtCompileTime> ret(1, x.cols());
-  for (size_type i = 0; i < x.cols(); i++) {
-    ret(i) = var(new internal::dot_self_vari(x.col(i)));
+template <typename Mat, require_rev_matrix_t<Mat>* = nullptr>
+inline auto columns_dot_self(const Mat& x) {
+  using T_return = promote_var_matrix_t<Eigen::RowVectorXd, Mat>;
+
+  arena_t<plain_type_t<Mat>> arena_x = x;
+  arena_t<Eigen::MatrixXd> arena_x_val = value_of(arena_x);
```

Comment:
`value_of(arena_x)` is (an expressions that evaluates to) `MatrixXd`, so `arena_t<decltype(value_of(arena_x))>` equals `arena_t<Eigen::MatrixXd>`

Reply:
Ah good call.

Another note though, does this make a copy for `var<matrix>` rn? I think after #2064 this will be better since then we use `arena_matrix` on the backend of `var_value<matrix>`